src/glang/codegen/builtin_callables.py

```diff
@@ -698,6 +700,198 @@ def assert_can_write_to(self) -> None:
+
+        (self._addr, self._to_write) = arguments
+        assert self._addr.result_type.storage_kind.is_reference()
+        assert not self._to_write.result_type.storage_kind.is_reference()
+
+        assert (
+            self._addr.result_type.convert_to_value_type() == self._to_write.result_type
+        )
+
+        super().__init__(
+            VoidType(),
+            Type.Kind.VALUE,
+            meta,
+        )
+
+    def __repr__(self) -> str:
+        return f"VolatileWriteExpression({self._to_write} to {self._addr})"
+
+    def generate_ir(self, ctx: IRContext) -> IROutput:
+        # Remove any indirect references
+        conv_to_write, extra_exprs_to_write = do_implicit_conversion(
+            self._to_write, self._to_write.result_type
+        )
+
+        conv_addr, extra_exprs_addr = do_implicit_conversion(
+            self._addr, self._addr.result_type
+        )
+
+        ir = IROutput()
+        ir.extend(self.expand_ir(extra_exprs_to_write, ctx))
+        ir.extend(self.expand_ir(extra_exprs_addr, ctx))
+
+        dbg = self.add_di_location(ctx, ir)
+
+        # store volatile <ty> <value>, ptr <pointer>
+        ir.lines.append(
+            f"store volatile {conv_to_write.ir_ref_with_type_annotation}, "
+            f" {conv_addr.ir_ref_with_type_annotation}, {dbg}",
+        )
+        return ir
+
+    @property
+    def ir_ref_without_type_annotation(self) -> str:
+        assert False
+
+    def assert_can_read_from(self) -> None:
+        raise OperandError("Cannot read from `__builtin_volatile_write()`")
+
+    def assert_can_write_to(self) -> None:
+        raise OperandError("Cannot assign to `__builtin_volatile_write()`")
+
+
+class VolatileReadExpression(BuiltinCallable):
+    def __init__(
+        self,
+        specialization: list[SpecializationItem],
+        arguments: list[TypedExpression],
+        meta: Meta,
+    ) -> None:
+        assert len(specialization) == 0
+        assert len(arguments) == 1
+
+        self.result_ref = None
+
+        (self._addr,) = arguments
+        assert self._addr.result_type.storage_kind.is_reference()
+
+        super().__init__(
+            self._addr.result_type.convert_to_value_type(),
+            Type.Kind.VALUE,
+            meta,
+        )
+
+    def __repr__(self) -> str:
+        return f"VolatileReadExpression({self._to_write} to {self._addr})"
+
+    def generate_ir(self, ctx: IRContext) -> IROutput:
+        # Remove any indirect references
+        conv_addr, extra_exprs_addr = do_implicit_conversion(
+            self._addr, self._addr.result_type
+        )
+
+        ir = self.expand_ir(extra_exprs_addr, ctx)
+
+        self.result_ref = f"%{ctx.next_reg()}"
+
+        dbg = self.add_di_location(ctx, ir)
+
+        # <result> = load volatile <ty> <value>, ptr <pointer>
+        ir.lines.append(
+            f"{self.result_ref} = load volatile {self.result_type.ir_type}, "
+            f"{conv_addr.ir_ref_with_type_annotation}, {dbg}",
```

Comment:
```suggestion
        # <result> = load volatile <ty>, ptr <pointer>
        ir.lines.append(
            f"{self.result_ref} = load volatile {self.result_type.ir_type}, "
            f"{conv_addr.ir_ref_with_type_annotation}, {dbg}",
```
Something doesn't look right here